### ISSUE TITLE
[Enhancement]Optimize volcano end-to-end scheduling large-scale pod performance

### ISSUE THREAD

**Issue by JesseStutler** [volcano-sh/volcano#3852]
### What is the problem you're trying to solve

Recently, I used the scripts in benchmark/sh to compare the performance of volcano and kube-scheduler under benchmark. There are several problems that I found needed to be solved

### Describe the solution you'd like

- Traffic conflict optimization. The metric called `apiserver_current_inflight_requests` of kube-apiserver is significantly higher when volcano schedules large-scale pods than when kube-scheduler schedules, which may cause pressure on kube-apiserver and cause network conflicts, affecting scheduling performance. After reviewing the codes, I found there two places causing the problem:
    - [ ] At the end of each session, the session will update queue status's allocated field. If the session interval is short, there will be a lot of queue status update requests. And the session will also update the nums of pods in `Running/Failed/Succeed` states in podgroup status. If these fields of podgroups only used in kubectl to show the nums of pods in each state, we may rewrite the logic , not to persist these nums fields in podgroups, like this pr:  https://github.com/volcano-sh/volcano/pull/3751
    - [ ] If we open gang plugin, gang plugin will always refresh the Scheduled condition at the end of each session(Refresh the transition ID), it's a bug need to be fixed. In the logic of `isPodGroupStatusUpdated`, because the nums of pod in Running state is always refreshing in scheduling, therefore `!equality.Semantic.DeepEqual(newStatus, oldStatus)` is true, causing the condition always will be refreshed: https://github.com/volcano-sh/volcano/blob/7170cca5a0af67938a05f46f3bf5095a127b13b9/pkg/scheduler/framework/job_updater.go#L81-L88 We also need to integrate Unschedulable and Scheduled conditions into the same condition, and use different ConditionStatus to distinguish them. Otherwise, even if the pods in the podgroups have met the minmember, the Unschedulable condition will still remain in the conditions list of the podgroup. , which would cause ambiguity. 
- Performance bottleneck of volcano admission. During the test, I found that we choose to use deployment to create large-scale pods, volcano-admission will seriously affect the speed of kube-controller-manager to create pods. We may need to rewrite or further optimize the logic of validate and mutate. 
    - [ ] validate webhook improvement( use CEL? )
    - [ ] mutate webhook improvement
- Deployment pods can not use the feature of predicateCache, the annotation of `volcano.sh/template-uid` is only patched on vc-job pods, predicateCache will firstly judge whether the annotation exists and then use the feature of predicateCache. And currently PredicateCache will only be used when predicateByStablefilter fails. Can we just use PredicateCache directly when pod template is the same? (Exclude some predicates such as podAntiAffinity and topologySpread)
    - [ ] Deployment pod uses the feature of predicateCache
    - [ ] Predicate cache enhancement?

In addition, we may need to reconstruct the logic of volcano's predicate and prioritize. Assume that there are m nodes in the cluster and n pods need to be scheduled. Currently, all m nodes need to be filtered. The time complexity is O(mn). Can we not filter all the nodes? Or after predicates some nodes, there is no need to filter all these nodes when scoring. 
- [ ] We can use heap or B-tree to quickly find the best node (but we need a uniform function to build heaps or B-trees, volcano has too many plugins, it may be complicated to find a uniform function)

**Comment by Monokaix** [comment]
Some previous performance issus: https://github.com/volcano-sh/volcano/issues/3502

**Comment by jxs1211** [comment]
Hi, newbie to volcano community, I'd like to make my first contribution to get my feet wet  in the community, and I'm looking at the first one, which hopes change the logic in the `UpdatePodGroup` of defaultStatusUpdater to avoid update the nums status in podgroup, like the [queueStatusApply](https://github.com/volcano-sh/volcano/pull/3751#issuecomment-2401202004) did, is that right?

**Comment by jxs1211** [comment]
`And currently PredicateCache will only be used when predicateByStablefilter fails. Can we just use PredicateCache directly when pod template is the same? (Exclude some predicates such as podAntiAffinity and topologySpread)`

I didn't find the logic of `PredicateCache will only be used when predicateByStablefilter fails`
I can only find the checking result of the PredicateWithCache, the cache will be updated if the the template-uid doesn't exist or there is no cache for the node and pod

```go
		predicateCacheStatus := make([]*api.Status, 0)
		if predicate.cacheEnable {
			fit, err = pCache.PredicateWithCache(node.Name, task.Pod)
			if err != nil {
				predicateCacheStatus, fit, _ = predicateByStablefilter(nodeInfo)
				pCache.UpdateCache(node.Name, task.Pod, fit)
			} else {
				if !fit {
					err = fmt.Errorf("plugin equivalence cache predicates failed")
					predicateCacheStatus = append(predicateCacheStatus, &api.Status{
						Code: api.Error, Reason: err.Error(), Plugin: CachePredicate,
					})
				}
			}
		} else {
			predicateCacheStatus, fit, _ = predicateByStablefilter(nodeInfo)
		}

		predicateStatus = append(predicateStatus, predicateCacheStatus...)
		if !fit {
			return api.NewFitErrWithStatus(task, node, predicateStatus...)
		}
```
Basd on the code, could you pls explain how can I use PredicateCache if the podTemplate is the same instead of checking the predicateByStablefilter fails?

**Comment by JesseStutler** [comment]
> Hi, newbie to volcano community, I'd like to make my first contribution to get my feet wet in the community, and I'm looking at the first one, which hopes change the logic in the `UpdatePodGroup` of defaultStatusUpdater to avoid update the nums status in podgroup, like the [queueStatusApply](https://github.com/volcano-sh/volcano/pull/3751#issuecomment-2401202004) did, is that right?

Yes, you can try this task! You can open a new issue and refer here when you finish your work.

**Comment by JesseStutler** [comment]
> And currently PredicateCache will only be used when predicateByStablefilter fails. Can we just use PredicateCache directly when pod template is the same? (Exclude some predicates such as podAntiAffinity and topologySpread)

Sorry it was my fault, ignore this. But prediacte only works for vc-job pod now, you can implement it to work for deployment also.

**Comment by JesseStutler** [comment]
> `And currently PredicateCache will only be used when predicateByStablefilter fails. Can we just use PredicateCache directly when pod template is the same? (Exclude some predicates such as podAntiAffinity and topologySpread)`

Sorry it was my fault, ignore this. But prediacte only works for vc-job pod now, you can implement it to work for deployment also.

**Comment by jxs1211** [comment]
> > `And currently PredicateCache will only be used when predicateByStablefilter fails. Can we just use PredicateCache directly when pod template is the same? (Exclude some predicates such as podAntiAffinity and topologySpread)`
> 
> Sorry it was my fault, ignore this. But prediacte only works for vc-job pod now, you can implement it to work for deployment also.

Is there any related or similar code can help me understand how to do and what to do, like statefulset implementation or something?

**Comment by JesseStutler** [comment]
> > > `And currently PredicateCache will only be used when predicateByStablefilter fails. Can we just use PredicateCache directly when pod template is the same? (Exclude some predicates such as podAntiAffinity and topologySpread)`
> > 
> > 
> > Sorry it was my fault, ignore this. But prediacte only works for vc-job pod now, you can implement it to work for deployment also.
> 
> Is there any related or similar code can help me understand how to do and what to do, like statefulset implementation or something?

No, you can check how the vc-job pods `"volcano.sh/template-uid"` are annotated in the controller and do some refactoring to implement deployment.

**Comment by lishangyuzi** [comment]
In the scenario of scheduling large-scale jobs, I also encountered a problem. When the job fails to be scheduled, all the pods under this job will update the PodCondition. Since it is necessary to communicate with the apiserver, this will take a long time.Could we consider using the multi-goroutine approach to handle this part of the logic?
https://github.com/volcano-sh/volcano/blob/0966fd59a9e0c0cfa7f6a19d41b3263f932ba4e0/pkg/scheduler/cache/cache.go#L1488

**Comment by lowang-bh** [comment]
> * volcano.sh/template-uid

I think we can use taskGroupID("%s/%s", task.Job, task.TaskRole) to unify those cases. 
There is already a taskSpec (from `volcano.sh/task-spec` annotation) in each role-pod。 Deployment also can use this annotation。

**Comment by Bnjoroge1** [comment]
Hi, is this issue still unaddressed? would love to take a stab at it

**Comment by Bnjoroge1** [comment]
I'll tentatively assign it to myself /assign

**Comment by JesseStutler** [comment]
> Hi, is this issue still unaddressed? would love to take a stab at it

Yes, you can choose some items to work on, especially the optimization of webhooks. Currently, webhook performance is a significant bottleneck, which affects the performance of batch scheduling. It would be best to move the webhook's validation or mutation to the kube-apiserver, either by modifying the CRD YAML or using validation/mutation admission policies, or simply by optimizing the architecture/performance of the webhooks

**Comment by AmanKumar1411** [comment]
Hi, is this issue still open ? i would like to work on this 
@Monokaix @JesseStutler

**Comment by zhuangzhewei09** [comment]
I’d like to work on the remaining predicateCache support for Deployment-created Pods.

My understanding is:
- predicateCache currently depends on `volcano.sh/template-uid`;
- vc-job Pods get this annotation, but Deployment/ReplicaSet Pods do not;
- as a result, Deployment workloads cannot benefit from predicateCache.

I plan to first verify:
1. where `volcano.sh/template-uid` is generated and consumed;
2. whether ReplicaSet UID / PodTemplateHash / ownerReferences can safely represent the same template identity;
3. whether adding the annotation in admission mutate path is acceptable, or whether predicateCache should derive template identity without mutating Pods.

I’ll start with a small proposal before opening a PR.

/assign

@JesseStutler
